### PR TITLE
State serialization improvements

### PR DESCRIPTION
**Pull request by dsrbecky**
Changes needed to serialize state for mid-execution-capture.

This code path is still mostly unused.

### CODE REVIEW

**Comment by AWoloszyn** [comment]
I don't see anything wrong here.

**Comment by AWoloszyn** [comment]
I don't see anything wrong here.

File: gapis/api/templates/api.proto.tmpl [hunk 1]
```diff
@@ -28,7 +28,7 @@
   {{end}}
 
   {{range $c := $.Classes}}
-    {{if (GetAnnotation $c "serialize")}}{{Template "ClassEntry" $c}}{{end}}
+    {{if not (GetAnnotation $c "noserialize")}}{{Template "ClassEntry" $c}}{{end}}
```
Comment (ben-clayton) [inline gapis/api/templates/api.proto.tmpl]: If you've killed everything that responds to the `@serialize` annotation, please can you remove the uses of the annotation?

File: gapii/cc/pack_encoder.cpp [hunk 2]
```diff
@@ -159,13 +150,15 @@ uint64_t PackEncoderImpl::writeTypeIfNew(const Descriptor* desc) {
...
     if (!type_id_cache) {
+        std::lock_guard<std::mutex> lock(mShared->mutex);
         return writeTypeIfNewBlocking(desc);
     }
 
     uint64_t type_id;
 
     auto iter_id = type_id_cache->type_ids.find(desc);
     if (iter_id == end(type_id_cache->type_ids)) {
+        std::lock_guard<std::mutex> lock(mShared->mutex);
```
Comment (ben-clayton) [inline gapii/cc/pack_encoder.cpp]: Any reason you move the lock outside of `writeTypeIfNewBlocking`?
It just seems more likely now that someone might call it without a lock.

Comment (dsrbecky) [reply gapii/cc/pack_encoder.cpp]: The lock is non-recursive, I can change it to recursive one, which is slightly bigger diff.

Comment (ben-clayton) [reply gapii/cc/pack_encoder.cpp]: Fair enough.

File: core/data/deep/copy_test.go [hunk 3]
```diff
@@ -98,7 +108,7 @@ func TestCopy(t *testing.T) {
 		err := deep.Copy(test.dst, test.src)
 		if assert.For(ctx, "err").ThatError(err).Succeeded() {
 			got := reflect.ValueOf(test.dst).Elem().Interface()
-			assert.For(ctx, "res").That(got).DeepEquals(test.expect)
+			assert.For(ctx, fmt.Sprintf("Test %v", i)).That(got).DeepEquals(test.expect)
```
Comment (ben-clayton) [inline core/data/deep/copy_test.go]: No need for the Sprintf, For is already a format printer.

File: gapis/api/templates/api.go.tmpl [hunk 4]
```diff
@@ -430,6 +430,15 @@
     }
   {{end}}
 
+  // Assign sets all fileds of this slive from different slice. Needed for cloning.
```
Comment (ben-clayton) [inline gapis/api/templates/api.go.tmpl]: s/fileds/fields
s/slive/slice

**Pull request merged**